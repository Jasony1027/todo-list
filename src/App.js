import React, { useState, useEffect } from "react";
import { nanoid } from "nanoid";
import Todo from "./components/Todo";
import Form from "./components/Form";
import FilterButton from "./components/FilterButton";

export const FILTER_MAP = {
  All: () => true,
  NotStarted: (task) => task.status === 0,
  Started: (task) => task.status === 1,
  Completed : (task) => task.status === 2
};
export const FILTER_NAMES = Object.keys(FILTER_MAP);

function App(props) {
  //hooks
  const [tasks, setTasks] = useState([]);
  const [filter, setFilter] = useState("All");
  useEffect(() => {
    const data = localStorage.getItem("listOfTasks");
    if (data) {
      setTasks(JSON.parse(data));
    }
  }, []);
  useEffect(() => {
    localStorage.setItem("listOfTasks", JSON.stringify(tasks));
  }, [tasks]);
  //lists
  const taskList = tasks
    .filter((task) => FILTER_MAP[filter](task))
    .map((task) => (
      <Todo
        id={task.id}
        name={task.name}
        status={task.status}
        key={task.id}
        toggleTaskStatus={toggleTaskStatus}
        deleteTask={deleteTask}
        editTask={editTask}
      />
    ));
  const filterList = FILTER_NAMES.map((name) => (
    <FilterButton
      key={name}
      name={name}
      isPressed={name === filter}
      setFilter={setFilter}
    />
  ));
  //functions
  function addTask(name) {
    const newTask = { id: "todo-" + nanoid(), name: name, status: 0 };
    setTasks([...tasks, newTask]);
    return name;
  }
    function toggleTaskStatus(id, status) {
    const updatedTasks = tasks.map((task) => {
      if (id === task.id) {

        return { ...task, status: status };
      }
      return task;
    });
    setTasks(updatedTasks);
  }
  function deleteTask(id) {
    const remainingTasks = tasks.filter((task) => id !== task.id);
    setTasks(remainingTasks);
  }
  function editTask(id, newName) {
    const editedTaskList = tasks.map((task) => {
      // if this task has the same ID as the edited task
      if (id === task.id) {
        return { ...task, name: newName };
      }
      return task;
    });
    setTasks(editedTaskList);
  }
  function clearTasks(){
    localStorage.clear();
    setTasks([]); 
  }
  //constants
  const tasksNoun = taskList.length !== 1 ? "tasks" : "task";
  const headingText = `${taskList.length} ${tasksNoun} remaining`;
  return (
    <div className="todoapp stack-large" data-testid="app">
      <h1>TodoMatic</h1>
      <Form addTask={addTask} />
      <div className="filters btn-group stack-exception">{filterList}</div>
      <h2 data-testid="headingText" id="list-heading">
        {headingText}
      </h2>
      <ul
        className="todo-list stack-large stack-exception"
        aria-labelledby="list-heading"
      >
        {taskList}
      </ul>
      <button type="button" className="btn btn__clearAll"  onClick={() => clearTasks()}>Clear All Tasks</button>
    </div>
  );
}

export default App;
