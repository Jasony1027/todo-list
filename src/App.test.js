import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import App, { FILTER_MAP, FILTER_NAMES } from "./App";

const mockData = [
  { id: "todo-0", name: "Code", completed: true },
  { id: "todo-1", name: "Sleep", completed: false },
  { id: "todo-2", name: "Repeat", completed: false },
];

afterEach(cleanup);

describe("The App component should be rendered successfully", () => {
  it("App component rendered without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App tasks={mockData} />, div);
  });

  it("App component snapshot test", () => {
    const componet = renderer.create(<App tasks={mockData} />);
    const tree = componet.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("FILTER_MAP object and the array FILTER_NAMES that contains its keys test", () => {
  it("FILTER_MAP object should contain 3 keys: All, Active, Completed", () => {
    expect(FILTER_MAP).toHaveProperty("All");
    expect(FILTER_MAP).toHaveProperty("Active");
    expect(FILTER_MAP).toHaveProperty("Completed");
  });
  it("FILTER_NAMES array should contain the keys of FILTER_MAP", () => {
    expect(FILTER_NAMES).toEqual(
      expect.arrayContaining(Object.keys(FILTER_MAP))
    );
  });
});

describe("Heading text test", () => {
  it("Heading text should display the number of active task(s)", () => {
    const { getByTestId } = render(<App tasks={mockData} />);
    const headingText = getByTestId("headingText");
    expect(headingText.textContent).toBe("0 tasks remaining");
  });
});
