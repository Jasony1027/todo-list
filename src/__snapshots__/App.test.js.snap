// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The App component should be rendered successfully App component snapshot test 1`] = `
<div
  className="todoapp stack-large"
  data-testid="app"
>
  <h1>
    TodoMatic
  </h1>
  <form
    onSubmit={[Function]}
  >
    <h2
      className="label-wrapper"
    >
      <label
        className="label__lg"
        htmlFor="new-todo-input"
      >
        What needs to be done?
      </label>
    </h2>
    <input
      autoComplete="off"
      className="input input__lg"
      id="new-todo-input"
      name="text"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      className="btn btn__primary btn__lg"
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    className="filters btn-group stack-exception"
  >
    <button
      aria-pressed="true"
      className="btn toggle-btn"
      onClick={[Function]}
      type="button"
    >
      <span
        className="visually-hidden"
      >
        Show 
      </span>
      <span>
        All
      </span>
      <span
        className="visually-hidden"
      >
         tasks
      </span>
    </button>
    <button
      aria-pressed="true"
      className="btn toggle-btn"
      onClick={[Function]}
      type="button"
    >
      <span
        className="visually-hidden"
      >
        Show 
      </span>
      <span>
        Active
      </span>
      <span
        className="visually-hidden"
      >
         tasks
      </span>
    </button>
    <button
      aria-pressed="true"
      className="btn toggle-btn"
      onClick={[Function]}
      type="button"
    >
      <span
        className="visually-hidden"
      >
        Show 
      </span>
      <span>
        Completed
      </span>
      <span
        className="visually-hidden"
      >
         tasks
      </span>
    </button>
  </div>
  <h2
    data-testid="headingText"
    id="list-heading"
  >
    0 tasks remaining
  </h2>
  <ul
    aria-labelledby="list-heading"
    className="todo-list stack-large stack-exception"
  />
  <button
    className="btn btn__clearAll"
    onClick={[Function]}
    type="button"
  >
    Clear All Tasks
  </button>
</div>
`;
